File: packages/superwall/android/src/main/java/io/capawesome/capacitorjs/plugins/superwall/classes/options/ConfigureOptions.java
package io.capawesome.capacitorjs.plugins.superwall.classes.options;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.getcapacitor.JSObject;
import com.getcapacitor.PluginCall;
import com.superwall.sdk.config.options.PaywallOptions;
import com.superwall.sdk.config.options.SuperwallOptions;
import com.superwall.sdk.logger.LogLevel;
import com.superwall.sdk.logger.LogScope;
import io.capawesome.capacitorjs.plugins.superwall.classes.CustomExceptions;
import java.util.EnumSet;
import org.json.JSONArray;

public class ConfigureOptions {

    @NonNull
    private final String apiKey;

    @Nullable
    private final SuperwallOptions options;

    public ConfigureOptions(@NonNull PluginCall call) throws Exception {
        this.apiKey = getApiKeyFromCall(call);
        this.options = buildSuperwallOptions(getOptionsFromCall(call));
    }

    @NonNull
    public String getApiKey() {
        return apiKey;
    }

    @Nullable
    public SuperwallOptions getSuperwallOptions() {
        return options;
    }

    @Nullable
    private static SuperwallOptions buildSuperwallOptions(@Nullable JSObject optionsObject) {
        if (optionsObject == null) {
            return null;
        }

        SuperwallOptions superwallOptions = new SuperwallOptions();

        // Paywall options
        JSObject paywallsObj = optionsObject.getJSObject("paywalls");
        if (paywallsObj != null) {
            PaywallOptions paywallOptions = new PaywallOptions();

            Boolean isHapticFeedbackEnabled = paywallsObj.getBool("isHapticFeedbackEnabled");
            if (isHapticFeedbackEnabled != null) {
                paywallOptions.setHapticFeedbackEnabled(isHapticFeedbackEnabled);
            }

            Boolean shouldShowPurchaseFailureAlert = paywallsObj.getBool("shouldShowPurchaseFailureAlert");
            if (shouldShowPurchaseFailureAlert != null) {
                paywallOptions.setShouldShowPurchaseFailureAlert(shouldShowPurchaseFailureAlert);
            }

            Boolean shouldPreload = paywallsObj.getBool("shouldPreload");
            if (shouldPreload != null) {
                paywallOptions.setShouldPreload(shouldPreload);
            }

            Boolean automaticallyDismiss = paywallsObj.getBool("automaticallyDismiss");
            if (automaticallyDismiss != null) {
                paywallOptions.setAutomaticallyDismiss(automaticallyDismiss);
            }

            superwallOptions.setPaywalls(paywallOptions);
        }

        // Logging options
        JSObject loggingObj = optionsObject.getJSObject("logging");
        if (loggingObj != null) {
            SuperwallOptions.Logging logging = new SuperwallOptions.Logging();

            String levelStr = loggingObj.getString("level");
            if (levelStr != null) {
                switch (levelStr) {
                    case "DEBUG":
                        logging.setLevel(LogLevel.debug);
                        break;
                    case "INFO":
                        logging.setLevel(LogLevel.info);
                        break;
                    case "WARN":
                        logging.setLevel(LogLevel.warn);
                        break;
                    case "ERROR":
                        logging.setLevel(LogLevel.error);
                        break;
                    default:
                        logging.setLevel(LogLevel.warn);
                        break;
                }
            }

            try {
                JSONArray scopesArray = loggingObj.getJSONArray("scopes");
                if (scopesArray != null && scopesArray.length() > 0) {
                    EnumSet<LogScope> scopes = EnumSet.noneOf(LogScope.class);
                    for (int i = 0; i < scopesArray.length(); i++) {
                        String scopeStr = scopesArray.getString(i);
                        LogScope scope = null;
                        switch (scopeStr) {
                            case "ALL":
                                scope = LogScope.all;
                                break;
                            case "CONFIG":
                                scope = LogScope.configManager;
                                break;
                            case "EVENTS":
                                scope = LogScope.paywallEvents;
                                break;
                            case "PAYWALLS":
                                scope = LogScope.paywallPresentation;
                                break;
                            case "PURCHASES":
                                scope = LogScope.productsManager;
                                break;
                        }
                        if (scope != null) {
                            scopes.add(scope);
                        }
                    }
                    if (!scopes.isEmpty()) {
                        logging.setScopes(scopes);
                    }
                }
            } catch (Exception e) {
                // Skip scopes if array is invalid
            }

            superwallOptions.setLogging(logging);
        }

        // Network environment
        String networkEnvStr = optionsObject.getString("networkEnvironment");
        if (networkEnvStr != null) {
            switch (networkEnvStr) {
                case "RELEASE":
                    superwallOptions.setNetworkEnvironment(new SuperwallOptions.NetworkEnvironment.Release());
                    break;
                case "DEVELOPER":
                    superwallOptions.setNetworkEnvironment(new SuperwallOptions.NetworkEnvironment.Developer());
                    break;
                default:
                    superwallOptions.setNetworkEnvironment(new SuperwallOptions.NetworkEnvironment.Release());
                    break;
            }
        }

        // Locale identifier
        String localeIdentifier = optionsObject.getString("localeIdentifier");
        if (localeIdentifier != null) {
            superwallOptions.setLocaleIdentifier(localeIdentifier);
        }

        // Android-specific options
        Boolean shouldObservePurchases = optionsObject.getBool("shouldObservePurchases");
        if (shouldObservePurchases != null) {
            superwallOptions.setShouldObservePurchases(shouldObservePurchases);
        }

        Boolean isExternalDataCollectionEnabled = optionsObject.getBool("isExternalDataCollectionEnabled");
        if (isExternalDataCollectionEnabled != null) {
            superwallOptions.setExternalDataCollectionEnabled(isExternalDataCollectionEnabled);
        }

        return superwallOptions;
    }

    @NonNull
    private static String getApiKeyFromCall(@NonNull PluginCall call) throws Exception {
        String apiKey = call.getString("apiKey");
        if (apiKey == null) {
            throw CustomExceptions.API_KEY_MISSING;
        }
        return apiKey;
    }

    @Nullable
    private static JSObject getOptionsFromCall(@NonNull PluginCall call) {
        return call.getObject("options");
    }
}

File: packages/superwall/ios/Plugin/Classes/Options/ConfigureOptions.swift
import Foundation
import Capacitor
import SuperwallKit

@objc public class ConfigureOptions: NSObject {
    let apiKey: String
    let options: SuperwallOptions?

    init(_ call: CAPPluginCall) throws {
        self.apiKey = try ConfigureOptions.getApiKeyFromCall(call)
        self.options = ConfigureOptions.buildSuperwallOptions(ConfigureOptions.getOptionsFromCall(call))
    }

    private static func getApiKeyFromCall(_ call: CAPPluginCall) throws -> String {
        guard let apiKey = call.getString("apiKey") else {
            throw CustomError.apiKeyMissing
        }
        return apiKey
    }

    private static func getOptionsFromCall(_ call: CAPPluginCall) -> JSObject? {
        return call.getObject("options")
    }

    private static func buildSuperwallOptions(_ optionsObject: JSObject?) -> SuperwallOptions? {
        guard let optionsObject = optionsObject else {
            return nil
        }

        let superwallOptions = SuperwallOptions()

        // Paywall options
        if let paywallsObj = optionsObject["paywalls"] as? JSObject {
            if let isHapticFeedbackEnabled = paywallsObj["isHapticFeedbackEnabled"] as? Bool {
                superwallOptions.paywalls.isHapticFeedbackEnabled = isHapticFeedbackEnabled
            }

            if let shouldShowPurchaseFailureAlert = paywallsObj["shouldShowPurchaseFailureAlert"] as? Bool {
                superwallOptions.paywalls.shouldShowPurchaseFailureAlert = shouldShowPurchaseFailureAlert
            }

            if let shouldPreload = paywallsObj["shouldPreload"] as? Bool {
                superwallOptions.paywalls.shouldPreload = shouldPreload
            }

            if let automaticallyDismiss = paywallsObj["automaticallyDismiss"] as? Bool {
                superwallOptions.paywalls.automaticallyDismiss = automaticallyDismiss
            }
        }

        // Logging options
        if let loggingObj = optionsObject["logging"] as? JSObject {
            if let levelStr = loggingObj["level"] as? String {
                switch levelStr {
                case "DEBUG":
                    superwallOptions.logging.level = .debug
                case "INFO":
                    superwallOptions.logging.level = .info
                case "WARN":
                    superwallOptions.logging.level = .warn
                case "ERROR":
                    superwallOptions.logging.level = .error
                default:
                    superwallOptions.logging.level = .warn
                }
            }

            if let scopesArray = loggingObj["scopes"] as? [String], !scopesArray.isEmpty {
                var scopes: Set<LogScope> = []
                for scopeStr in scopesArray {
                    switch scopeStr {
                    case "ALL":
                        scopes.insert(.all)
                    case "ANALYTICS":
                        scopes.insert(.analytics)
                    case "CONFIG":
                        scopes.insert(.configManager)
                    case "EVENTS":
                        scopes.insert(.paywallEvents)
                    case "PAYWALLS":
                        scopes.insert(.paywallPresentation)
                    case "PURCHASES":
                        scopes.insert(.productsManager)
                    default:
                        break
                    }
                }
                if !scopes.isEmpty {
                    superwallOptions.logging.scopes = scopes
                }
            }
        }

        // Network environment
        if let networkEnvStr = optionsObject["networkEnvironment"] as? String {
            switch networkEnvStr {
            case "RELEASE":
                superwallOptions.networkEnvironment = .release
            case "DEVELOPER":
                superwallOptions.networkEnvironment = .developer
            default:
                superwallOptions.networkEnvironment = .release
            }
        }

        // Locale identifier
        if let localeIdentifier = optionsObject["localeIdentifier"] as? String {
            superwallOptions.localeIdentifier = localeIdentifier
        }

        return superwallOptions
    }
}
